### PR TITLE
Replace Mutex with RWMutex increase read efficiency

### PR DESCRIPTION
**Pull request by wgliang**
Replace Metux with RWMutex increase read efficiency.

### CODE REVIEW

**Comment by irfanurrehman** [comment]
@shivramsrivastava, do you have any tests already setup which can verify changes like these? If not, any plan ??

**Comment by shivramsrivastava** [comment]
@irfanurrehman
Yes, we have some tests, but as of now, we are in the process of setting the verify scripts and the CI  scripts for this project.


@wgliang 
There are many changes not present in this repo as of now.
We will be merging them soon after we have the CI scripts.
This change which you have posted was already addressed in here
https://github.com/shivramsrivastava/poseidon/blob/dev/pkg/k8sclient/podwatcher.go

Please bear with us, while we refresh this repo with all the latest changes.

**Comment by wgliang** [comment]
@shivramsrivastava  Ok, that's all right.

**Comment by dhilipkumars** [comment]
@shivramsrivastava yes could we hold of on other PRs until we set up the initial verify stage for this product, please? plus these PRs will inturn help us verify if our script actually work as expected or not. 

+ you seem posses more power than any other approver, your lgtms are considered as approved. :-) :-)

**Comment by ICGog** [comment]
\ltgm

However, let's not push this until we have the testing pipeline in place.

**Comment by shivramsrivastava** [comment]
@dhilipkumars I was looking into the prow commands list.
Below is the link to the list of commands for our repo.
https://prow.k8s.io/command-help?repo=kubernetes-sigs%2Fscheduling_poseidon

It looks like a user with approver privilege both 'lgtm' and 'approve' has the same effect.

I don't know how it works with other repo?
Do we need to raise an issue for this?

**Comment by wgliang** [comment]
@ICGog  ltgm or lgtm？

File: pkg/k8sclient/types.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 const bytesToKb = 1024
 
 // Used to guard access to the pod, task and job related maps.
-var PodsCond *sync.Cond
+var PodsMux *sync.RWMutex
```
Comment (dhilipkumars) [inline pkg/k8sclient/types.go]: @shivramsrivastava 
Shouldn't `PodMux` be a part of `PodWatcher` struct instead? I don't see this being used anywhere else. 

Plus what about other maps below? can they also be moved inside `PodWatcher` ?

```
 var PodToTD map[PodIdentifier]*firmament.TaskDescriptor
 var TaskIDToPod map[uint64]PodIdentifier
 var jobIDToJD map[string]*firmament.JobDescriptor
```

File: pkg/k8sclient/podwatcher.go [hunk 2]
```diff
@@ -258,7 +258,7 @@ func (this *PodWatcher) podWorker() {
 				switch pod.State {
 				case PodPending:
 					glog.V(2).Info("PodPending ", pod.Identifier)
-					PodsCond.L.Lock()
+					PodsMux.Lock()
```
Comment (dhilipkumars) [inline pkg/k8sclient/podwatcher.go]: the name `PodMux` is ambiguous with `Pod Multiplexer`, could we simply call `podMutex` to be clearer?

File: pkg/k8sclient/types.go (hunk 1 above)
Comment (shivramsrivastava) [reply pkg/k8sclient/types.go]: @dhilipkumars 
The maps and the mutexes are accessed across packages like stats package, k8sclient, and the main package. So they are global.

The stats package use these global mutexes to lock and read/update the pods and nodes from the already populated maps. If the map doesn't have an entry for the pod or node the stats package will not update the pod or node info.

 **About stats module:**
    Very briefly stats module gets the info from heaspter and filters what is required and send that information to the firmament's knowledgebase.

So we had kept them as global.

> I will think of something to make them closer to where they belong, for clarity.

File: pkg/k8sclient/podwatcher.go (hunk 2 above)
Comment (wgliang) [reply pkg/k8sclient/podwatcher.go]: You're right, I will modify it.

**Comment by wgliang** [comment]
@irfanurrehman @shashidharatd PTAL,Thx.

**Comment by wgliang** [comment]
@dhilipkumars

**Comment by m1093782566** [comment]
@wgliang Please figure the verify out - probably you need to rebase the HEAD.

Once all CI checks are green, I will lgtm this PR.

**Comment by wgliang** [comment]
/cc m1093782566 
Please help merge!

**Comment by wgliang** [comment]
/cc @m1093782566 
Please help merge!

**Comment by m1093782566** [comment]
Okay, let's soak it in test earlier.

/lgtm

**Pull request merged**